### ISSUE TITLE
Bug: In v8, no-unused-vars says "defined but only used as type" on import with no escape hatch

### ISSUE THREAD

**Issue by jakebailey** [typescript-eslint/typescript-eslint#9679]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

With code like:

```ts
import { SomeType } from "./someModule";

export const value = 1234 as typeof SomeType;
```

This code is flagged as having an unused variable in v8 with a new message, something like:

```
/home/jabaile/work/TypeScript/src/harness/client.ts
  53:5  warning  'PatternMatchKind' is defined but only used as a type. Allowed unused vars must match /^(_+$|_[^_])/u  @typescript-eslint/no-unused-vars
```

The TS repo canonically does not use type-only imports (we cannot use `verbatimModuleSyntax`), so it's not desirable to make this a type-only import. But, there's no way to opt out of this new detection for imports.

### Reproduction Repository Link

https://github.com/jakebailey/TypeScript

### Repro Steps

1) Clone https://github.com/jakebailey/TypeScript/tree/eslint-9
2) `npm ci`
3) `npx eslint src/harness/client.ts`

```
/home/jabaile/work/TypeScript/src/harness/client.ts
  53:5  warning  'PatternMatchKind' is defined but only used as a type. Allowed unused vars must match /^(_+$|_[^_])/u  @typescript-eslint/no-unused-vars
```

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `8.0.0-alpha.59` |
| `@typescript-eslint/parser`            | `8.0.0-alpha.59` |
| `@typescript-eslint/scope-manager`     | `8.0.0-alpha.59` |
| `@typescript-eslint/typescript-estree` | `8.0.0-alpha.59` |
| `@typescript-eslint/type-utils`        | `8.0.0-alpha.59` |
| `@typescript-eslint/utils`             | `8.0.0-alpha.59` |
| `TypeScript`                           | `5.4.4` |
| `ESLint`                               | `9.8.0` |
| `node`                                 | `22.5.1` |

**Comment by kirkwaiblinger** [comment]
Hi @jakebailey!

Hmm, looks like this was an intentional change added recently in https://github.com/typescript-eslint/typescript-eslint/issues/3325 / https://github.com/typescript-eslint/typescript-eslint/pull/9330

Note that the vars ignore pattern does appear to work as an escape hatch ([playground link](https://v8--typescript-eslint.netlify.app/play/#ts=5.4.3&fileType=.tsx&code=JYWwDg9gTgLgBAbwFBzgZQiApgFQJ5hZwCGAznAPobb6FIC%2BcAZlJnAEQB0A9KZlgFkIAEwCuAGyzsA3EiRYAHpFhwAxhAB2peADdi40UQC8cAIwAmAMwAWEuRgEsEJpWq5H0oA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tidmUQAmtAG4BDaKgwBtcNhwB3cU0gAaMLLlY52SGIkBJAOadoiAAqj8%2BRNGUZIAPQD6AOgBUkDdgC%2BXgLqyvt5AA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false))

Would that work for your use case, or would you like to advocate for a different solution?

**Comment by jakebailey** [comment]
That doesn't really seem like a good way to ignore the problem... If we then go to use that import (it's an enum), we'd need to rename it back.

To me, `no-unused-var` is overstepping; this is already enforced by `consistent-type-imports` as switching it to `type SomeType` would also fix the problem ([Playground Link](https://v8--typescript-eslint.netlify.app/play/#ts=5.4.3&fileType=.tsx&code=JYWwDg9gTgLgBAbwFBzjAnmApnAyhELAFUyyQF84AzKAuAIgDoB6AZwKwFkIATAVwA2WegG4kSLAA9IsOAGMIAO1bwAbgEMBfHAF44ARgBMAZgAscdazSkIVPBxLYRQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tidmUQAmtAG4BDaKgwBtcNhwB3cU0gAaMLLlY52SGIkBJAOadoiAAqj8%2BRNGUZIAPQD6AOgBUkDdgC%2BXgLqyvt5AA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)).

**Comment by jakebailey** [comment]
Of course, my comment mainly only applies to imports; #3325 is talking about am unreferenced local.

**Comment by Josh-Cena** [comment]
This is not overstepping though; this new report would also flag a local value declaration that's only used in `typeof`.

**Comment by jakebailey** [comment]
Yes, see my previous comment (and that is largely why I titled the issue in relation to imports).

**Comment by kirkwaiblinger** [comment]
Yeah I'm leaning towards @jakebailey's side on this, that some special-casing behavior around imports makes sense here.

1. The current behavior, as @jakebailey pointed out, is an edge-case of #3325, not the motivation for the change in the first place. 
2. Again, as @jakebailey pointed out, dealing with type-only imports is its whole own consideration, by separate rules and compiler options. It feels needlessly unergonomic to prohibit values used in `typeof` in a codebase which doesn't use `import type` at all, whereas `import`ing types used as types would be allowed

**Comment by bradzacher** [comment]
This is an oversight -- #9330 should not have applied to imports!
Imports are "dual value/type" variables in the scope analysis and so they should not be reported on by the `typeof` check.

**Comment by bradzacher** [comment]
Also here is a [repro against the v8 playground](https://v8--typescript-eslint.netlify.app/play/#ts=5.5.4&showAST=scope&fileType=.tsx&code=JYWwDg9gTgLgBAbzgZQiApgFQJ5nXAXzgDMo04AiAOgHoBnNdAWQgBMBXAG3QoG4AofugAekWHADGEAHZ14ANwCGndvgC8cAIwAmAMwAWOIrpwYudBGIpGOPAKA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tidmUQAmtAG4BDaKgyRE0aB2iRwYAL4hlQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

**Comment by jakebailey** [comment]
Sent #9694.

**Comment by doberkofler** [comment]
I have a related use case that in my code base causes the "warning  'z$foo' is assigned a value but only used as a type  @typescript-eslint/no-unused-vars" error when for example using the zod library. I get this error every time a zod definition it already used to infer the type but (not yet) to parse.
I my code baseI get this error a few hundred times and it would be most helpful to have an option that exclude this specific  use case of unused variable.

```typescript
const z$foo = z.object({bar: z.boolean()});
export type foo = z.infer<typeof z$foo>;
```

[playground](https://v8--typescript-eslint.netlify.app/play/#ts=5.5.2&fileType=.tsx&code=JYWwDg9gTgLgBAbzgLzgXzgMyhEcDkyEAJvgNwBQFAxhAHYDO8yAJJhBHALwoB0EAIwBWAU2owAFEgEBDKAC4%2BAjgBsRMuhICU6LZREAPSLDgwAnmBFYO3PsDqYRUADznLETCjYcAfJQpAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

**Comment by doberkofler** [comment]
Good question!
We actually generate the zod schema objects from an external definition and always export the inferred type.
The definition itself might or might not be used to parse untyped external data.

**Comment by vinniespina** [comment]
i commit
